### PR TITLE
Persist the datadir after option reset

### PR DESCRIPTION
**Pull request by achow101**
Clicking the "Reset Options" button in "Settings" will remove all of the settings, including any custom data directory that the user sets during the first start of the software. This PR persists the datadir setting following option reset since setting the datadir afterwards can be difficult for the average user given that the only way to set the datadir after the first startup is to use the `--datadir` option.

### CODE REVIEW

**Comment by jonasschnelli** [comment]
Concept ACK.
Issue with this PR: If the user passes `-resetguisettings`, I think we should also evict the custom set data dir.

**Comment by MarcoFalke** [comment]
> If the user passes -resetguisettings, I think we should also evict the custom set data dir.

Which raises the question if `-resetguisettings` should cause the intro to pop up again (force `-choosedatadir`).

**Comment by achow101** [comment]
> Issue with this PR: If the user passes -resetguisettings, I think we should also evict the custom set data dir.

Why should it? The primary issue is that the user doesn't know the datadir has been reset. So then resetting the options should either keep the datadir or have the user choose the datadir again.

**Comment by laanwj** [comment]
> Which raises the question if -resetguisettings should cause the intro to pop up again (force -choosedatadir).

I think it should either reset the data directory *and* reask with the dialog at start, or keep the data directory the same. The middle ground makes no sense.

**Comment by jonasschnelli** [comment]
> I think it should either reset the data directory and reask with the dialog at start, or keep the data directory the same. The middle ground makes no sense.

Agree and I think it should re-ask with the init dialog at start.

**Comment by achow101** [comment]
Ok, I will change this PR to reask for the datadir at start.

**Comment by achow101** [comment]
I've updated it so that it will reask for the datadir at the start. I've also left the datadir persist stuff since it is still needed for if `-resetguisettings` is used.

**Comment by laanwj** [comment]
Separate from the changes here, we have a bug in this logic. There are two user actions that can result in a settings reset:

- A1: Start bitcoin-qt with `-resetguisettings`. This results in `app.createOptionsModel(true)`, which calls `OptionsModel::Reset()`, before going into `OptionsModel::Init`.
- A2: Go to the options dialog and click the "Reset options" button. This calls `model->Reset()` and quits the application.

A2 is okay. A1 however, happens *after* it is no longer legal to change the data directory! (this should happen before step 6 in `main`)

We need to move the `-resetguisettings` logic *before* `Intro::pickDataDirectory()`. This will make sure that A1 and A2 act in the same, sane, way.

**Comment by laanwj** [comment]
Oh, shit, that won't work either.

- Step 7 determines the network (testnet or mainnet)
- And we want `-resetguisettings` to clear the options for the current network, not the mainnet ones, so doing it before step 6 would clear the wrong settings.

This initialization sequence is so damn complex because it is possible to put `testnet=1` in the `bitcoin.conf` file, as this file is in the mainnet directory even for testnet/regtest... so it can override the *real* data directory used later.

**Comment by achow101** [comment]
Yeah, so to deal with A1, I left in the code that keeps the datadir. So what happens is that if the user does anything to reset the options, the choose datadir window will appear and let them choose the datadir. Then, if A1 happens, the datadir is still set and will remain the same after the options are reset.

**Comment by laanwj** [comment]
Ok, makes sense.

Another question: what is the specific need for `fReset`? What makes a reset state different from just a clean slate? Both should get you into the datadir choose dialog, assuming the default datadir doesn't exist.

**Comment by achow101** [comment]
A reset state does not remove the datadir. A user may be using the default 
datadir and reset the options. We want the behavior to be consistent 
regardless of the datadir.

Also, it is possible that the default datadir could already exist, perhaps 
from moving the datadir but not deleting it, or testnet or regtest used the 
default datadir but mainnet does not. This would mean that it finds the 
datadir and thus doesn't show the choose dialog.

File: src/qt/intro.cpp [hunk 1]
```diff
@@ -204,6 +204,7 @@ void Intro::pickDataDirectory()
         }
 
         settings.setValue("strDataDir", dataDir);
+	settings.setValue("fReset", false);
```
Comment (laanwj) [inline src/qt/intro.cpp]: nit: wrong indentation

**Comment by jonasschnelli** [comment]
Tested ACK https://github.com/bitcoin/bitcoin/tree/57b85246d313ac5af94cc86ed98ae4a665e40800 modulo indentation nit.

**Comment by laanwj** [comment]
@achow101 this seems ready to merge apart from the indentation nit ,can you please fix that?

**Comment by achow101** [comment]
Nit addressed

**Comment by jonasschnelli** [comment]
Tested ACK 15df3c1

**Comment by luke-jr** [comment]
Backport needed?

**Comment by achow101** [comment]
> Backport needed?

Is it?

**Pull request merged**